x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexing/AsyncTwoPhaseIndexer.java

```diff
@@ -158,7 +158,13 @@ public synchronized boolean maybeTriggerAsyncJob(long now) {
                         if (r) {
                             nextSearch(ActionListener.wrap(this::onSearchResponse, this::finishWithSearchFailure));
                         } else {
-                            finishAndSetState();
+                            // If our state is INDEXING then transition to STARTED
+                            // Calling checkState here covers the scenarios when
+                            // the indexer was aborted or set to STOPPING between `onStart` being called
+                            // and this branch being executed
+                            if (checkState(getState())) {
```

Comment:
I don't understand why we need this check since `finishAndSetState` checks the current state and choose the appropriate transition. Is the issue that `finishAndSetState` doesn't save the state ? Should we change the code to:
`doSaveState(finishAndSetState(), getPosition(), () -> {})` ?

Reply:
@jimczi I updated this, please take a second look. 

There is definitely a race condition here that is painful. The issue with calling a `doSaveState` always is that the implementer does not know if the transition was from `STOPPING -> STOPPED` or `STOPPED -> STOPPED`. Or if the indexer was aborted (in which case we probably don't want to save the state.